### PR TITLE
Adding RBAC and PSP by default

### PR DESCRIPTION
**Pull request by puja108**
towards https://github.com/giantswarm/giantswarm/issues/1331 and https://github.com/giantswarm/giantswarm/issues/1788

### CODE REVIEW

**Comment by puja108** [comment]
RFR

The flag, bindings and roles are what has been running for a few days now on the staging cluster on `asgard`.

Only additional thing here is the rbac bootstrap as a first step in the addons script. Applying the bindings first and then the roles worked out for me in my manual tests.

This needs to be tested for example on `gauss`, as my testing was limited to manually changing an already running cluster. Should work, but I'd rather see it coming up as a fresh cluster before rolling this out.

**Comment by puja108** [comment]
If this works we can do a similar PR for PodSecurityPolicies.

**Comment by xh3b4sd** [comment]
Talking about the dev side. This looks to me like we should have some addon operator and/or an permissions operator running inside the guest clusters. We will have to template all these things for AWS/KVM/Azure the same way so we should have one solution for this to not need to copy/paste a lot of crap in each and every operator for each and every provider. An operator would make sense to always ensure such resources are properly set up. We recently had an issue with a customer where the ingress controllers within a guest cluster had just be gone and nobody knew why. Having an operator taking care about these resources would also prevent such scenarios.

**Comment by JosephSalisbury** [comment]
I really have no if this is "right" or not :D

**Comment by puja108** [comment]
fuck, I forgot the templating...next commit incoming

**Comment by puja108** [comment]
@xh3b4sd you are right, this we should do in the future, actually we should just not use our own addon shit and go with the community addon-manager, which already has reconcilliation loops. I actually already prepared some of our deployments for this (see annotations). But this is a different issue that I was hoping to avoid by moving to `kubeadm` but we shelved that so...

**Comment by puja108** [comment]
should be alright now, anyone know how we can test this? @rossf7 usually does that part for my PRs, so I'm kinda clueless. I just know there's some vendoring dance to do with the operators

**Comment by puja108** [comment]
I tested podsecuritypolicies together with this and decided to add this to the PR so we only test once and then get 2 security features that should be used together anyway with one PR

**Comment by puja108** [comment]
PSPs are now added, with this setup only cluster-admin role can run privileged containers (and mount privileged directories like hostpath).

Next:
I will add roles for the 2 PSPs. This way a cluster admin can flexibly assign privileged access. I will add a binding that gives all authenticated users restricted access only.

**Comment by puja108** [comment]
Done, now need someone to help me with testing or tell me how to do it.

**Comment by puja108** [comment]
Running (manually set up) on `56x77` on `ginger`

**Comment by puja108** [comment]
nearly there, I changed the upstream restricted PSP to be a little bit less restricted (allowing users in containers running as root) as it otherwise blocks nearly anything users want to deploy if its not customized to be more secure. If customers ask for stricter policy they can create another one.

doing some final testing now

**Comment by puja108** [comment]
final test came up on `gauss` cluster `z6apr`

**Comment by puja108** [comment]
conformance test ran through without degradation (still 124/125)

Can someone LGTM this? you can check above cluster on gauss to see if it is running.

ping @calvix @teemow at some point we should introduce that in host cluster, too.

**Comment by puja108** [comment]
Next I will write some docs for this I guess

File: templates.go [hunk 1]
```diff
@@ -1049,7 +1467,6 @@ coreos:
       ExecStartPre=/usr/bin/docker pull $IMAGE
       ExecStartPre=-/usr/bin/docker stop -t 10 $NAME
       ExecStartPre=-/usr/bin/docker rm -f $NAME
-      ExecStartPre=/bin/sh -c "while ! curl --output /dev/null --silent --head --fail --cacert /etc/kubernetes/ssl/apiserver-ca.pem --cert /etc/kubernetes/ssl/apiserver-crt.pem --key /etc/kubernetes/ssl/apiserver-key.pem https://{{.Cluster.Kubernetes.API.Domain}}; do sleep 1 && echo 'Waiting for master'; done"
```
Comment (teemow) [inline templates.go]: This looks like it doesn't belong into this PR. 

Why did you remove this?

File: templates.go [hunk 2]
```diff
@@ -1488,10 +1906,6 @@ coreos:
       ExecStartPre=/usr/bin/docker pull $IMAGE
       ExecStartPre=-/usr/bin/docker stop -t 10 $NAME
       ExecStartPre=-/usr/bin/docker rm -f $NAME
-      ExecStartPre=/bin/bash -c "while [ ! -f /etc/kubernetes/ssl/worker-ca.pem ]; do echo 'Waiting for /etc/kubernetes/ssl/worker-ca.pem to be written' && sleep 1; done"
```
Comment (teemow) [inline templates.go]: Same here: https://github.com/giantswarm/k8scloudconfig/pull/180/files#r136763850

**Comment by teemow** [comment]
Are the docs already updated? I think this shouldn't be merged until we know all our examples work with RBAC enabled.

File: templates.go (hunk 1 above)
Comment (puja108) [reply templates.go]: We actually removed this from all units long time ago, seems we forgot some units. I needed to remove this for RBAC/PSP as this would prevent stuff from coming up. We are not running into race conditions without it, the components just retry, so it's not needed.

**Comment by puja108** [comment]
@teemow we have updates for the xamples already prepared for RBAC, I'll check on the rest.

**Comment by puja108** [comment]
Docs TODO:

- [ ] update examples in docs-content (https://github.com/giantswarm/docs-content/pull/255)
- [ ] update examples outside docs-content (prom & elastic open)
- [ ] add docs for usage re usr mgmt

**Comment by puja108** [comment]
Can be merged after https://github.com/giantswarm/kubernetes-elastic-stack/pull/34 and https://github.com/giantswarm/docs-content/pull/257

**Comment by puja108** [comment]
Can I get an LGTM?

**Comment by calvix** [comment]
also merge master before updating

**Comment by puja108** [comment]
docs PRs are RFR

**Comment by puja108** [comment]
waiting for #1798

**Comment by jgsqware** [comment]
@puja108  Regarding https://github.com/giantswarm/giantswarm/issues/1719, **2.11** expect authorization plugins to validate. I think RBAC and PSP enforce it too. WDYT?

**Comment by puja108** [comment]
closing in favor of https://github.com/giantswarm/k8scloudconfig/pull/203
